`RandomizedRaftTest#consistencyTestWithSnapshot` fails with uncaught exception

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? Once so far: https://github.com/camunda/zeebe/actions/runs/6491028769/job/17627701526
- Does it block your work? No
- Do we suspect that it is a real failure? Possibly

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
java.lang.AssertionError: Uncaught exception
	at org.junit.Assert.fail(Assert.java:89)
	at io.atomix.raft.DeterministicSingleThreadContext$WrappedRunnable.run(DeterministicSingleThreadContext.java:112)
	at org.jmock.lib.concurrent.DeterministicScheduler$CallableRunnableAdapter.call(DeterministicScheduler.java:176)
	at org.jmock.lib.concurrent.DeterministicScheduler$ScheduledTask.run(DeterministicScheduler.java:251)
	at org.jmock.lib.concurrent.DeterministicScheduler.runNextPendingCommand(DeterministicScheduler.java:66)
	at org.jmock.lib.concurrent.DeterministicScheduler.runUntilIdle(DeterministicScheduler.java:56)
	at org.jmock.lib.concurrent.DeterministicScheduler.tick(DeterministicScheduler.java:45)
	at io.atomix.raft.ControllableRaftContexts.tick(ControllableRaftContexts.java:313)
	at io.atomix.raft.ControllableRaftContexts.tickElectionTimeout(ControllableRaftContexts.java:287)
	at io.atomix.raft.RaftOperation.run(RaftOperation.java:42)
	at io.atomix.raft.RandomizedRaftTest.consistencyTest(RandomizedRaftTest.java:180)
	at io.atomix.raft.RandomizedRaftTest.consistencyTestWithSnapshot(RandomizedRaftTest.java:96)
</pre>
</details>

**Hypotheses**
Not clear what happened. Unfortunately, the actual exception is lost because it is printed to stderr and not logged properly:
https://github.com/camunda/zeebe/blob/b7df1f5ee03d9187a1c15e640928be3aeb64f62d/atomix/cluster/src/test/java/io/atomix/raft/DeterministicSingleThreadContext.java#L106-L114

**Logs**

[randomized-raft-test.zip](https://github.com/camunda/zeebe/files/12878159/randomized-raft-test.zip)

--- a/atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java
+++ b/atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java
@@ -279,7 +279,7 @@ public class RaftContext implements AutoCloseable, HealthMonitorable {
   }
 
   private void onNewPersistedSnapshot(final PersistedSnapshot persistedSnapshot) {
-    threadContext.execute(() -> setCurrentSnapshot(persistedSnapshot));
+    threadContext.execute(this::updateCurrentSnapshot);
   }
 
   private void onUncaughtException(final Throwable error) {
@@ -1256,9 +1256,12 @@ public class RaftContext implements AutoCloseable, HealthMonitorable {
     return currentSnapshot;
   }
 
-  public void setCurrentSnapshot(final PersistedSnapshot snapshot) {
+  public void updateCurrentSnapshot() {
     checkThread();
-    currentSnapshot = snapshot;
+    // Get the latest snapshot from snapshot store because it might have been updated already before
+    // this listener is executed
+    currentSnapshot = persistedSnapshotStore.getLatestSnapshot().orElse(null);
+    log.trace("Set currentSnapshot to {}", currentSnapshot);
     logCompactor.compactFromSnapshots(persistedSnapshotStore);
   }
 

--- a/atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java
+++ b/atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java
@@ -259,7 +259,7 @@ public class PassiveRole extends InactiveRole {
       pendingSnapshotStartTimestamp = 0L;
       snapshotReplicationMetrics.decrementCount();
       snapshotReplicationMetrics.observeDuration(elapsed);
-      raft.setCurrentSnapshot(persistedSnapshot);
+      raft.updateCurrentSnapshot();
       onSnapshotReceiveCompletedOrAborted();
     } else {
       setNextExpected(request.nextChunkId());

--- a/atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java
+++ b/atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java
@@ -48,6 +48,7 @@ import io.camunda.zeebe.journal.JournalException.InvalidChecksum;
 import io.camunda.zeebe.journal.JournalException.InvalidIndex;
 import io.camunda.zeebe.snapshots.PersistedSnapshot;
 import io.camunda.zeebe.snapshots.ReceivedSnapshot;
+import io.camunda.zeebe.snapshots.SnapshotException.SnapshotAlreadyExistsException;
 import java.nio.ByteBuffer;
 import java.util.concurrent.CompletableFuture;
 import org.agrona.concurrent.UnsafeBuffer;
@@ -187,8 +188,16 @@ public class PassiveRole extends InactiveRole {
                     .build()));
       }
 
-      pendingSnapshot =
-          raft.getPersistedSnapshotStore().newReceivedSnapshot(snapshotChunk.getSnapshotId());
+      try {
+        pendingSnapshot =
+            raft.getPersistedSnapshotStore().newReceivedSnapshot(snapshotChunk.getSnapshotId());
+      } catch (final SnapshotAlreadyExistsException snapshotAlreadyExists) {
+        // This should not happen because we previously check for the latest snapshot. But, if it
+        // happens, instead of crashing raft thread, we respond with success because we already have
+        // the snapshot.
+        return CompletableFuture.completedFuture(
+            logResponse(InstallResponse.builder().withStatus(RaftResponse.Status.OK).build()));
+      }
       log.info("Started receiving new snapshot {} from {}", pendingSnapshot, request.leader());
       pendingSnapshotStartTimestamp = System.currentTimeMillis();
       snapshotReplicationMetrics.incrementCount();

--- a/atomix/cluster/src/test/java/io/atomix/raft/DeterministicSingleThreadContext.java
+++ b/atomix/cluster/src/test/java/io/atomix/raft/DeterministicSingleThreadContext.java
@@ -24,8 +24,13 @@ import java.time.Duration;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
 import org.jmock.lib.concurrent.DeterministicScheduler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public final class DeterministicSingleThreadContext implements ThreadContext {
+
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(DeterministicSingleThreadContext.class);
 
   private final DeterministicScheduler deterministicScheduler;
 
@@ -108,8 +113,8 @@ public final class DeterministicSingleThreadContext implements ThreadContext {
       try {
         command.run();
       } catch (final Exception e) {
-        e.printStackTrace();
-        fail("Uncaught exception");
+        LOGGER.error("Uncaught exception", e);
+        fail("Uncaught exception" + e);
       }
     }
   }

--- a/atomix/cluster/src/test/java/io/atomix/raft/snapshot/InMemorySnapshot.java
+++ b/atomix/cluster/src/test/java/io/atomix/raft/snapshot/InMemorySnapshot.java
@@ -254,4 +254,19 @@ public final class InMemorySnapshot implements PersistedSnapshot, ReceivedSnapsh
         && id.equals(that.id)
         && chunks.equals(that.chunks);
   }
+
+  @Override
+  public String toString() {
+    return "InMemorySnapshot{"
+        + "index="
+        + index
+        + ", term="
+        + term
+        + ", id='"
+        + id
+        + '\''
+        + ", checksum="
+        + checksum
+        + '}';
+  }
 }

